src/declare.rs

```diff
@@ -237,6 +244,87 @@ impl Drop for ClassDecl {
     }
 }
 
+/// Describes a method of an Objective-C protocol.
+pub trait MethodDescription {
+    /// The arguments to the protocol method.
+    type Args: EncodeArguments;
+
+    /// The return type of the protocol method.
+    type Ret: Encode;
+}
+
+impl<Args: EncodeArguments, Ret: Encode> MethodDescription for (Args, Ret) {
+    type Args = Args;
+    type Ret = Ret;
+}
+
+/// A type for declaring a new protocol and adding new methods to it
+/// before registering it.
+pub struct ProtocolDecl {
+    proto: *mut Protocol
+}
+
+impl ProtocolDecl {
+    /// Constructs a `ProtocolDecl` with the given name. Returns `None` if the
+    /// protocol couldn't be allocated.
+    pub fn new(name: &str) -> Option<ProtocolDecl> {
+        let c_name = CString::new(name).unwrap();
+        let proto = unsafe {
+            runtime::objc_allocateProtocol(c_name.as_ptr())
+        };
+        if proto.is_null() {
+            None
+        } else {
+            Some(ProtocolDecl { proto: proto })
+        }
+    }
+
+    fn add_method_description_common<M: MethodDescription>(&mut self, sel: Sel, is_required: bool,
```

Comment:
Hmm, since `MethodDescription` is always just a pair, I might recommend just passing the args and ret separately, and then it seems like the entire `MethodDescription` trait won't be needed. That sounds good because it's one less type to maintain! It also means one less pair of parens is needed, i.e. you can use `add_method<(Bar,), Foo>` instead of `add_method<((Bar,), Foo)>`. Maybe slightly more readable?

I was trying to see if I could suggest a more ergonomic API for passing the types, but I've got nothing.